### ISSUE TITLE
nextcloud logs remote ip-address as the gateways ip-address

### ISSUE THREAD

**Issue by szaimen** [nextcloud/all-in-one#2041]
We need to find a way to work around this. I am sure that it has worked at some point but seems like something changed and now it doesnt anymore. Can be easily verified by trying to log in with wrong credentials.

**Comment by szaimen** [comment]
@Zoey2936 could it be possible that this was cause by the apache migration to alpine?

**Comment by szaimen** [comment]
Looks like adding `"userland-proxy": false` to /etc/docker/daemon.json made it work...

**Comment by szaimen** [comment]
> is this only in non reverse proxy mode? since in /settings/admin/logging is see the correct ips

No, it is in default mode... Maybe it is due to the docker daemon update to 23 that I recently installed?

**Comment by Zoey2936** [comment]
I also use docker v23.0.1...

**Comment by szaimen** [comment]
That is very strange. but since it fixes the issue for me, I guess it is worth documenting.

**Comment by szaimen** [comment]
yeah, don't know why it works for you even without that setting. Anyhow, it is useful to document this imho.

**Comment by Zoey2936** [comment]
I know why it works for me...
We trust private IPs only in reverse proxy mode: https://github.com/nextcloud/all-in-one/blob/dfc2d9798ee69b0183b4ad0299ab2e187b76d436/Containers/apache/start.sh#L42

**Comment by szaimen** [comment]
> We trust private IPs only in reverse proxy mode:

I thought the same and tried this but it didnt made it work for me...

**Comment by szaimen** [comment]
The actual only thing that made it work was disabling the userland-proxy.

**Comment by szaimen** [comment]
I just tried that again and it didnt work without disabling the userland-proxy. Anyhow, it might come with a few downside so the wiki article must be enough for now.

**Comment by szaimen** [comment]
@Zoey2936 I just had the thought why it works for you: this issue does not appear if the container uses the network_mode host. Only if the container runs in a bridged network.

**Comment by szaimen** [comment]
So basically I still think that this is the way to go for a setup without reverse proxy. Reopening https://github.com/nextcloud/all-in-one/pull/2044

**Comment by Zoey2936** [comment]
I've set up a new AIO instance without reverse proxy, and it is working:
![grafik](https://user-images.githubusercontent.com/75573284/220406749-599f4054-fe65-4927-ad6c-c6b8443fc203.png)

**Comment by Zoey2936** [comment]
root@ip:~# docker version
Client: Docker Engine - Community
 Version:           23.0.1
 API version:       1.42
 Go version:        go1.19.5
 Git commit:        a5ee5b1
 Built:             Thu Feb  9 19:46:54 2023
 OS/Arch:           linux/amd64
 Context:           default

Server: Docker Engine - Community
 Engine:
  Version:          23.0.1
  API version:      1.42 (minimum version 1.12)
  Go version:       go1.19.5
  Git commit:       bc3805a
  Built:            Thu Feb  9 19:46:54 2023
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.6.18
  GitCommit:        2456e983eb9e37e47538f59ea18f2043c9a73640
 runc:
  Version:          1.1.4
  GitCommit:        v1.1.4-0-g5fd4c4d
 docker-init:
  Version:          0.19.0
  GitCommit:        de40ad0

**Comment by Zoey2936** [comment]
the daemon.json is empty

**Comment by szaimen** [comment]
Okay then I am really out of ideas now.

**Comment by Zoey2936** [comment]
> Okay then I am really out of ideas now.

whats your base OS?

**Comment by szaimen** [comment]
Ubuntu 22.04 lts

**Comment by Zoey2936** [comment]
My servers run with debian 11 (backports)

**Comment by szaimen** [comment]
Hm... I fear i dont have time to debug this further, I found a solution that works and documented it so its fine for me now. But thanks for your help!

**Comment by Zoey2936** [comment]
Can I ask what IP the logs exactly reported for you?

**Comment by szaimen** [comment]
It logged the gateway ip...

**Comment by Zoey2936** [comment]
did you try this with an ipv4 or ipv6 internet connection?

**Comment by szaimen** [comment]
Ipv4

**Comment by Zoey2936** [comment]
That was the last Idea I had, I have a IPv4 only internet connection and docker seems to have a lot of problems with IPv6... (https://github.com/moby/moby/issues/17666)

**Comment by Zoey2936** [comment]
I think even this sounds similar, this is not the problem: https://github.com/moby/moby/issues/25526

**Comment by szaimen** [comment]
One idea that I have that maybe based on how you install docker userland proxy might get enabled or not but no idea how to check this. So e.g. when you install docker from ubuntu repo it enables it and when installing via script from docker not...

**Comment by Zoey2936** [comment]
what is your output of this?
`docker info --format '{{.CgroupDriver}}' `

**Comment by szaimen** [comment]
systemd

**Comment by szaimen** [comment]
This is docker info on my system:
```
Client:
 Context:    default
 Debug Mode: false
 Plugins:
  buildx: Docker Buildx (Docker Inc.)
    Version:  v0.10.2
    Path:     /usr/libexec/docker/cli-plugins/docker-buildx
  compose: Docker Compose (Docker Inc.)
    Version:  v2.15.1
    Path:     /usr/libexec/docker/cli-plugins/docker-compose
  scan: Docker Scan (Docker Inc.)
    Version:  v0.23.0
    Path:     /usr/libexec/docker/cli-plugins/docker-scan

Server:
 Containers: 11
  Running: 8
  Paused: 0
  Stopped: 3
 Images: 26
 Server Version: 23.0.0
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Using metacopy: false
  Native Overlay Diff: true
  userxattr: false
 Logging Driver: json-file
 Cgroup Driver: systemd
 Cgroup Version: 2
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: io.containerd.runc.v2 runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 31aa4358a36870b21a992d3ad2bef29e1d693bec
 runc version: v1.1.4-0-g5fd4c4d
 init version: de40ad0
 Security Options:
  apparmor
  seccomp
   Profile: builtin
  cgroupns
 Kernel Version: 5.15.0-58-generic
 Operating System: Ubuntu 22.04.1 LTS
 OSType: linux
 Architecture: x86_64
 CPUs: 2
 Total Memory: 3.738GiB
 Name: aio-testing1
 ID: 5T4Z:Y64Q:ET2R:72X2:25DQ:4BU4:6NZ3:WGHF:6WTS:ON3V:YQPS:3UJZ
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 Registry: https://index.docker.io/v1/
 Experimental: false
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false
```

**Comment by Zoey2936** [comment]
My looks like this:
```
Client:
 Context:    default
 Debug Mode: false
 Plugins:
  buildx: Docker Buildx (Docker Inc.)
    Version:  v0.10.2
    Path:     /usr/libexec/docker/cli-plugins/docker-buildx
  compose: Docker Compose (Docker Inc.)
    Version:  v2.16.0
    Path:     /usr/libexec/docker/cli-plugins/docker-compose
  scan: Docker Scan (Docker Inc.)
    Version:  v0.23.0
    Path:     /usr/libexec/docker/cli-plugins/docker-scan

Server:
 Containers: 95
  Running: 92
  Paused: 0
  Stopped: 3
 Images: 76
 Server Version: 23.0.1
 Storage Driver: overlay2
  Backing Filesystem: extfs
  Supports d_type: true
  Using metacopy: false
  Native Overlay Diff: true
  userxattr: false
 Logging Driver: json-file
 Cgroup Driver: systemd
 Cgroup Version: 2
 Plugins:
  Volume: local
  Network: bridge host ipvlan macvlan null overlay
  Log: awslogs fluentd gcplogs gelf journald json-file local logentries splunk syslog
 Swarm: inactive
 Runtimes: io.containerd.runc.v2 runc
 Default Runtime: runc
 Init Binary: docker-init
 containerd version: 2456e983eb9e37e47538f59ea18f2043c9a73640
 runc version: v1.1.4-0-g5fd4c4d
 init version: de40ad0
 Security Options:
  apparmor
  seccomp
   Profile: builtin
  cgroupns
 Kernel Version: 6.0.0-0.deb11.6-amd64
 Operating System: Debian GNU/Linux 11 (bullseye)
 OSType: linux
 Architecture: x86_64
 CPUs: 8
 Total Memory: 23.48GiB
 Name: <hostname>
 ID: SNZ6:FZ6V:V6US:LUDL:KH5R:BDNE:J2B3:B7OI:CQKA:EPAJ:JJHM:4OBS
 Docker Root Dir: /var/lib/docker
 Debug Mode: false
 Username: <username>
 Registry: https://index.docker.io/v1/
 Experimental: true
 Insecure Registries:
  127.0.0.0/8
 Live Restore Enabled: false
```

**Comment by szaimen** [comment]
okay, it looks like my rourter updated itself and enabled ipv6 automatically. However ipv6 in docker is not enabled on my test server. So this combination might have triggered some strange behaviour? Anyway, disabling userland proxy fixed it for me reliably so I think this is still the way to go if one should experience this issue.

**Comment by Zoey2936** [comment]
and if you enable userland proxy and enable ipv6, with my daemon json from above, does this work? can you please try it?

**Comment by szaimen** [comment]
@Zoey2936 based on your report it looks like https://github.com/docker/docs/blob/main/config/daemon/ipv6.md is out-of-date? Would you mind opening a PR there to add your suggested config there?

**Comment by Zoey2936** [comment]
the config I used is recommended by mailcow and the differnc between the mailcow settings and docker settings is nat, mailcow creates a nat and only uses one public ip, while docker would give every container a public ip, which is most time a bad idea, since every port would be open to the internet

**Comment by Zoey2936** [comment]
but did you tried it with my deamon.json?

**Comment by szaimen** [comment]
> the config I used is recommended by mailcow and the differnc between the mailcow settings and docker settings is nat, mailcow creates a nat and only uses one public ip, while docker would give every container a public ip, which is most time a bad idea, since every port would be open to the internet

Ah I see. However what I meant was this: https://github.com/docker/docs/compare/main...szaimen:docs:patch-1. Does this look good to you?

> but did you tried it with my deamon.json?

No, not yet

**Comment by Zoey2936** [comment]
> Ah I see. However what I meant was this: [docker/docs@main...szaimen:docs:patch-1](https://github.com/docker/docs/compare/main...szaimen:docs:patch-1). Does this look good to you?

If I remember correctly, the experimental option has nothing to do with ipv6 directly and ip6tables make only sense in combination with fixed-cidr-v6 set to a private ip range

**Comment by szaimen** [comment]
I see, So both things are actually not needed or is ip6tables?

**Comment by szaimen** [comment]
I know now why disabling userland-proxy works: it basically disables all ipv6 features in docker:

with userland-proxy disabled:
![image](https://user-images.githubusercontent.com/42591237/235735432-33c2fb9c-f07b-44e6-8075-92dbfd14c1ea.png)

with userland-proxy enabled:
![image](https://user-images.githubusercontent.com/42591237/235735492-139a5251-ac38-490c-a318-7e7a5f9783e6.png)

**Comment by szaimen** [comment]
So basically it looks like not enabling ipv6 for the internal network will result in it logging the ip-address of the docker gateway instead of the actual ipv6-address.

**Comment by szaimen** [comment]
Yes, network_mode for the RP seems indeed to be the holy grail!

**Comment by szaimen** [comment]
So my currently plan would be to enable ipv6 for the nextcloud-aio network if ipv6 is enabled for the default bridge network. we would need to retreive the ip-address form that so that it does not overlap with it. same for the ipv4...
